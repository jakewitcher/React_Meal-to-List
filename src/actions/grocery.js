export const onAddGroceryAsync = (
    {
        name = '',
        items = [],
    } = {}
) => ({
    type: 'ADD_GROCERY_ASYNC',
    grocery: {
        name,
        items,
    }
});

export const onDeleteGroceryAsync = (id = '') => ({
    type: 'DELETE_GROCERY_ASYNC',
    id,
});

export const onEditGroceryAsync = (
    {
        name = '',
        id = '',
        items = [],
    } = {}
) => ({
    type: 'EDIT_GROCERY_ASYNC',
    grocery: {
        id,
        name,
        items,
    }
});

export const onSetGroceryAsync = () => ({
    type: 'SET_GROCERY_ASYNC',
});