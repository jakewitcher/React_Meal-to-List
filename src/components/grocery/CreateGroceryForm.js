import React from 'react';

const CreateGroceryForm = (props) => {
    const {
        selectedMealChange,
        selectedMeal,
        meals,
        groceryName,
        itemsList,
        groceryNameChange,
        addMealToList,
        resetGrocery,
        dispatch,
    } = props.groceryProps;
    const { addGrocery } = props;
    return (
        <div className="form">
            <h2 className="form__header">Create a New Grocery List</h2>
            <div className="form__form">
                <div className="form__nameinputfield">
                    <p>Grocery List Name</p>
                    <input
                        className="form__nameinputfield--input"
                        type="text"
                        value={groceryName}
                        onChange={groceryNameChange}
                    />
                </div>
                <div className="form__iteminputfields">
                    <div>
                        <p>Meals</p>
                        <select
                            className="form__iteminputfields--input form__iteminputfields--dropdown"
                            name="selectedMeal"
                            value={selectedMeal.name}
                            onChange={selectedMealChange}
                        >
                            {
                                meals.mealList.map(meal => <option value={meal.name} key={meal.id}>{meal.name}</option>)
                            }
                        </select>
                    </div>
                    <div className="form__button">
                        <button
                            className="form__button--add"
                            onClick={addMealToList}
                        >
                            Add Meal
                    </button>
                    </div>
                </div>
                <div>
                    <button
                        className="form__button--save"
                        onClick={() => {
                            dispatch(addGrocery({
                                name: groceryName,
                                items: itemsList,
                            }));
                            resetGrocery();
                        }
                        }
                    >
                        Save Grocery List
                </button>
                </div>
            </div>

        </div>
    )
}

export default CreateGroceryForm;