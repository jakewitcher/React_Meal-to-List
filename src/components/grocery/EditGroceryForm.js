import React from 'react';

const EditGroceryForm = (props) => {
    const {
        selectedMealChange,
        selectedMeal,
        meals,
        groceryName,
        itemsList,
        groceryNameChange,
        addMealToList,
        resetGrocery,
        dispatch,
    } = props.groceryProps;
    const { editGrocery, match } = props;
    return (
        <div className="form">
            <h2 className="form__header">Edit Grocery List</h2>
            <div className="form__form">
                <div className="form__nameinputfield">
                    <p>Grocery List Name</p>
                    <input
                        className="form__nameinputfield--input"
                        type="text"
                        value={groceryName}
                        onChange={groceryNameChange}
                    />
                </div>
                <div className="form__iteminputfields">
                    <div>
                        <p>Meals</p>
                        <select
                            className="form__iteminputfields--input form__iteminputfields--dropdown"
                            name="selectedMeal"
                            value={selectedMeal.name}
                            onChange={selectedMealChange}
                        >
                            {
                                meals.mealList.map(meal => <option value={meal.name} key={meal.id}>{meal.name}</option>)
                            }
                        </select>
                    </div>
                    <div className="form__button">
                        <button
                            className="form__button--add"
                            onClick={addMealToList}
                        >
                            Add Meal
                    </button>
                    </div>
                </div>
                <div>
                    <button
                        className="form__button--save"
                        onClick={() => {
                            const id = match.params.groceryId;
                            dispatch(editGrocery({
                                name: groceryName,
                                items: itemsList,
                                id,
                            }));
                            resetGrocery();
                        }
                        }
                    >
                        Save Grocery List
                </button>
                </div>
            </div>

        </div>
    )
}

export default EditGroceryForm;