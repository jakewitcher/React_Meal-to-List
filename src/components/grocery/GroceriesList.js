import React from 'react';
import Grocery from './Grocery';

const GroceriesList = (props) => {
    const { groceryLists } = props;
    return (
        <div>
            {
                groceryLists.map(list => {
                    return <Grocery
                        name={list.name}
                        key={list.id}
                    />
                })
            }
        </div>
    );
}

export default GroceriesList;