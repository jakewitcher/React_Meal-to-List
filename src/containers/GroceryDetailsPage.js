import React from "react";
import { connect } from "react-redux";
import GroceryItems from "../components/grocery/GroceryItems";

export const GroceryDetailsPage = props => {
  return (
    <div className="details__container">
      <div className="details">
        <GroceryItems
          groceryName={props.grocery.name}
          itemList={props.grocery.itemList}
          isDetails={true}
        />
      </div>
    </div>
  );
};

const mapStateToProps = (state, props) => {
  return {
    groceryLists: state.groceryLists,
    grocery: state.groceryLists.groceryList.filter(
      list => list.id === props.match.params.groceryId
    )[0]
  };
};

export default connect(mapStateToProps)(GroceryDetailsPage);
