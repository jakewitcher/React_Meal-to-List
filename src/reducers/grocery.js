const groceryDefaultState = {
    groceryList: [],
};

export default (state = groceryDefaultState, action) => {
    switch (action.type) {
        case 'ADD_GROCERY':
            return {
                ...state,
                groceryList: [...state.groceryList, action.grocery]
            };
        default:
            return state;
    }
}