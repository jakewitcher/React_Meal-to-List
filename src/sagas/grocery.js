import { put, takeEvery, select } from "redux-saga/effects";
import database from "../firebase/firebase";

const getUserId = state => state.auth.uid;

const formatItemList = list => {
  return list.reduce((list, item) => {
    list[item.itemName] = {
      amount: item.amount,
      unit: item.unit
    };
    return list;
  }, {});
};

function* addGrocery({ grocery = {} }) {
  const { itemList, name, groceryListMeals } = grocery;
  const items = formatItemList(itemList);
  const meals = groceryListMeals;

  let id;
  const uid = yield select(getUserId);

  yield database
    .ref(`users/${uid}/groceryLists`)
    .push({ name, items, meals })
    .then(ref => {
      id = ref.key;
    });

  yield put({
    type: "ADD_GROCERY",
    grocery: {
      name,
      itemList,
      groceryListMeals,
      id
    }
  });
}

function* editGrocery({ grocery = {} }) {
  const { itemList, name, id, groceryListMeals } = grocery;
  const items = formatItemList(itemList);
  const uid = yield select(getUserId);
  const meals = groceryListMeals;

  yield database
    .ref(`users/${uid}/groceryLists/${id}`)
    .update({ name, items, meals });

  yield put({
    type: "EDIT_GROCERY",
    grocery: {
      name,
      itemList,
      groceryListMeals,
      id
    }
  });
}

function* deleteGrocery({ id = "" }) {
  const uid = yield select(getUserId);

  yield database.ref(`users/${uid}/groceryLists/${id}`).remove();

  yield put({
    type: "DELETE_GROCERY",
    id
  });
}

function* setGrocery() {
  const lists = [];
  const uid = yield select(getUserId);

  yield database
    .ref(`users/${uid}/groceryLists`)
    .once("value")
    .then(snapshot => {
      snapshot.forEach(childSnapshot => {
        const itemList = [];
        childSnapshot.child("items").forEach(item => {
          itemList.push({
            itemName: item.key,
            amount: item.val().amount,
            unit: item.val().unit
          });
        });

        lists.push({
          id: childSnapshot.key,
          name: childSnapshot.val().name,
          itemList,
          groceryListMeals: childSnapshot.val().meals
        });
      });
    });
  yield put({ type: "SET_GROCERY", lists });
}

export function* watchAddGrocery() {
  yield takeEvery("ON_ADD_GROCERY", addGrocery);
}

export function* watchEditGrocery() {
  yield takeEvery("ON_EDIT_GROCERY", editGrocery);
}

export function* watchDeleteGrocery() {
  yield takeEvery("ON_DELETE_GROCERY", deleteGrocery);
}

export function* watchSetGrocery() {
  yield takeEvery("ON_SET_GROCERY", setGrocery);
}
