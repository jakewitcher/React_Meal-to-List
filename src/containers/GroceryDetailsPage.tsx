import React from "react";
import { connect } from "react-redux";
import GroceryItems from "../components/GroceryItems";
import { ApplicationState } from "../store/index";
import { RouteComponentProps } from "react-router";
import { Grocery } from "../store/grocery/types";

interface IGroceryDetailsPageProps
  extends RouteComponentProps<{ groceryId: string }> {
  grocery: Grocery;
}

export const GroceryDetailsPage: React.FC<IGroceryDetailsPageProps> = props => {
  return (
    <div className="details__container">
      <div className="details">
        <GroceryItems
          groceryName={props.grocery.name}
          itemList={props.grocery.itemList}
          isDetails={true}
        />
      </div>
    </div>
  );
};

const mapStateToProps = (
  state: ApplicationState,
  props: IGroceryDetailsPageProps
) => {
  return {
    groceryLists: state.groceryLists,
    grocery: state.groceryLists.groceryList.filter(
      list => list.id === props.match.params.groceryId
    )[0]
  };
};

export default connect(mapStateToProps)(GroceryDetailsPage);
