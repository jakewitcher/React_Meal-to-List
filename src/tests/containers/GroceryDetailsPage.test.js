import React from 'react';
import { shallow } from 'enzyme';
import { GroceryDetailsPage } from '../../containers/GroceryDetailsPage';
import { groceryLists } from '../fixtures/grocery';

test('should correctly render GroceryDetailsPage', () => {
    const grocery = groceryLists[0]
    const wrapper = shallow(<GroceryDetailsPage 
        grocery={grocery}
    />);
    expect(wrapper).toMatchSnapshot();
});