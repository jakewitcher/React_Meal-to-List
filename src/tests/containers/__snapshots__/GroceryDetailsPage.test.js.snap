// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render GroceryDetailsPage 1`] = `
<div>
  <GroceryItems
    groceryName="Monday Meals"
    isDetails={true}
  />
</div>
`;
